Objects/regions.c

```diff
@@ -670,8 +720,277 @@ PyObject* _Py_MakeImmutable(PyObject* obj)
+    /* The source of the reference that created the region error.
+     *
+     * A weak reference, can be made into a strong reference with `Py_INCREF`
+     */
+    PyObject* src;
+    /* The target of the reference that created the region error.
+     *
+     * A weak reference, can be made into a strong reference with `Py_INCREF`
+     */
+    PyObject* tgt;
+    /* This ID indicates what kind of error occurred.
+     */
+    region_error_id id;
+} regionerror;
+
+/* Used by `add_to_region` to handle errors. The first argument is
+ * the error information. The second argument is supplementary data
+ * passed along by `add_to_region`.
+ *
+ * The return indicates if `add_to_region` should continue adding
+ * other possible objects to the region.
+ */
+typedef int (*handle_add_to_region_error)(regionerror *, void *);
+
+/* This takes the region error and emits it as a `RegionError` to the
+ * user. This function will always return `false` to stop the propagation
+ * from `add_to_region`
+ *
+ * This function borrows both arguments. The memory has to be managed
+ * the caller.
+ */
+int emit_region_error(regionerror *error, void*) {
+    const char* msg = NULL;
+
+    switch (error->id)
+    {
+    case ERR_INTER_REGION_REF:
+        msg = "Inter region references are only allowed to bridge objects";
+        break;
+    case ERR_CYCLE_CREATION:
+        msg = "Regions are not allowed to create cycles";
+        break;
+    case ERR_SHARED_CUSTODY:
+        msg = "Regions can only have one parrent at a time";
+        break;
+    case ERR_WIP_FUNCTIONS:
+        msg = "WIP: Functions in regions are not supported yet";
+        break;
+    default:
+        assert(false && "unreachable?");
+        break;
+    }
+    throw_region_error(error->src, error->tgt, msg, NULL);
+
+    // We never want to continue once an error has been emitted.
+    return false;
+}
+
+typedef struct addtoregionvisitinfo {
+    stack* pending;
+    Py_uintptr_t region;
+} addtoregionvisitinfo;
+
+int _add_to_region_visit(PyObject* obj, void* info_void)
+{
+    addtoregionvisitinfo *info = _Py_CAST(addtoregionvisitinfo *, info_void);
+
+    // The object is already known and doesn't need to be traversed
+    if (Py_REGION(obj) == info->region || _Py_IsImmutable(obj)) {
+        return 0;
+    }
+
+    // The actual region update and write checks are done in the
+    // main body of `add_to_region`
+    if (stack_push(info->pending, obj)) {
+        PyErr_NoMemory();
+        return -1;
+    }
+
+    return 0;
+}
+
+/* This adds the given object and transitive objects to the given region.
+ * Errors will be passed to the given `handle_error` function along with
+ * the `handle_error_data`.
+ *
+ * The given region may contain flags. Objects from the local regions will
+ * have their region replaced with the `flagged_region` value. This will
+ * lose flags set on the local region.
+ */
+PyObject *add_to_region(PyObject *obj, Py_uintptr_t flagged_region,
```

Comment:
With "actual region" I assume you mean a `regionmetadata*`. For the cleaning code, I wanted to also set the "visited" flag. So this is a region which can contain flags, but it doesn't need to have flags set.

Reply:
Okay, could you document that behaviour and intent please.

Reply:
Flagged region suggests a property on the region being pointed to rather than a potentially tagged pointer to a region. I suggest some name changing here to clarity.  

On a related note, from experience with working with a code base needing to always distinguish between tagged and untagged pointers, it was a huge step forward when that started getting tracked in the types. So we might have a type for tagged pointer an a function that takes the tagged pointer to an untagged one (and back etc). That saved a lot of time scrolling up and down trying to figure out what kind of uintptr_t one was dealing with. :)